src/providers/ad/ad_gpo.c

```diff
@@ -4923,9 +4923,16 @@ ad_gpo_get_sd_referral_send(TALLOC_CTX *mem_ctx,
         goto done;
     }
 
+    state->smb_host = talloc_strdup(state, lud->lud_host);
+
```

Comment:
Hi,

if a user has specified a dedicated AD DC for this domain with the `ad_server` option letting `gpo_child` pick a random AD DC from this domain might not be what the user was expecting because e.g. the given AD DC is the only only reachable for the client.

There is a note in the `sssd-ad` man page "Note: Trusted domains will always auto-discover servers even if the primary server is explicitly defined in the ad_server option." which was added 2015 but imo is not true anymore.

If you would move the original code into `ad_gpo_get_sd_referral_conn_done()` the `uri` should be set.

bye,
Sumit

Reply:
This is indeed the right fix, I have tested it and `state->conn->service->uri` is set after `sdap_id_op_connect_{send,recv}`.